### ISSUE TITLE
Math block delimiter is the same for inline and display math

### ISSUE THREAD

**Issue by stsievert** [gettalong/kramdown#672]
Currently, Kramdown uses double dollar sign delimiters for inline *and* display math: https://kramdown.gettalong.org/syntax.html#math-blocks. I can make that work, but it's annoying because I have a bunch of blog posts written in Markdown that are processed through Pandoc. Pandoc's inline math delimiter is `$` and their display math delimiter is `$$` ([docs][pandoc]). If I want to use Kramdown, I have to convert all inline math delimiters to `$$` and display math delimiters to `\n$$...$$\n`. Converting will take a long time.

[pandoc]:https://pandoc.org/MANUAL.html#math

**Comment by stsievert** [comment]
Having the same delimiter for inline and display math is not standard:

* TeX uses `$` or `\(` for inline math and `$$` or `\[` for display math ([docs][tex]). This is perhaps the most important example; LaTeX equations were originally developed alongside TeX typesetting.
* Pandoc uses `$` for inline math and `$$` for display math ([docs][pandoc])
    * RMarkdown uses Pandoc, so they have the same delimiters.
* Jupyter Notebook uses `$` for inline math and `$$` for display math ([docs][jlab])
* Math StackExchange uses `$` for inline math and `$$` for display math ([docs][mse])
* MathJax uses `$$` for display math and `\(` for inline math ([docs][mj], which provides instructions on how to use `$` for inline math).
* KaTeX mirrors MathJax (relevant comment: https://github.com/KaTeX/KaTeX/issues/712#issuecomment-305283518). Their auto-render example they show how to use `$$` and `\[` for display math and `$` and `\(` for inline math.

In all the examples above "a delimiter of `\(`" means this Markdown is used to get inline math:

``` markdown
This is inline math: \(x^2 + x + 1 = 0\). This equation is often called the "golden equation."
```


[tex]:https://en.wikibooks.org/wiki/LaTeX/Mathematics#Mathematics_environments
[mj]:https://docs.mathjax.org/en/latest/input/tex/delimiters.html
[mse]:https://math.stackexchange.com/editing-help#link-latex
[pandoc]:https://pandoc.org/MANUAL.html#math
[jlab]:https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html#LaTeX-equations

**Comment by gettalong** [comment]
Using a single dollar sign may have undesired consequences for the mass of people who *don't* use math spans with kramdown. Therefore I decided to use the same syntax for math spans and blocks to avoid this problem. It is also a bit simpler because you only have to know one syntax.

If you need a different syntax for your math spans, you need to create a new subclass derived from the kramdown parser and adapt the math span parser regex and code.

Generally, I advise people to stick with one Markdown parser because documents written for one parser are nearly always not compatible with another parser. If you like Pandoc and have used it, best would be to stay with Pandoc and its Markdown syntax. If you want to switch to kramdown, you either need to adapt kramdown to your needs or rewrite your documents.

**Comment by stsievert** [comment]
> Using a single dollar sign may have undesired consequences for the mass of people who don't use math spans with kramdown. ... If you need a different syntax for your math spans, you need to create a new subclass derived from the kramdown parser and adapt the math span parser regex and code.

I agree: kramdown should not change by default.

However, I've never written a line of Ruby, and only use kramdown through Jekyll. Right now, my site bypasses kramdown math processing with `math_engine: nil`. I give the math processing to client-side JS through KaTeX because it supports the syntax common among LaTeX, Pandoc, Jupyter and Stack Exchange as mentioned above. I want to use kramdown's math processing because it can be performed server side

I think the transition from KaTeX/MathJax can be made easier by adding an optional parameter to support different math span delimiters. By default, it would not change the behavior of Kramdown. If enabled, it would allow math spans to be delimited by another character (e.g, `$`). This would better support naive Ruby users like me.

**Comment by gettalong** [comment]
Thanks for your thoughtful comment. I get what you are aiming for but I'm sorry, adding options for changing the parsing syntax is out of the question. There is only *one* kramdown syntax and it should stay this way.

If you want kramdown to process your math elements, you need to adapt your source documents. Best option is probably running a replacement script over your source files and then using `git diff` to spot invalid changes, i.e. where you really had a dollar sign and not a math start delimiter.

**Comment by sp301415** [comment]
Hi, I know this is an old & closed issue, but I just stumbled upon a same problem. I ended up writing my own parser:

```ruby
require 'kramdown/parser/kramdown'
require 'kramdown-parser-gfm'

class Kramdown::Parser::GFMKatex < Kramdown::Parser::GFM
    # Override inline math parser
    @@parsers.delete(:inline_math)

    INLINE_MATH_START = /(\$+)([^\$]+)(\$+)/m

    def parse_inline_math
        start_line_number = @src.current_line_number
        @src.pos += @src.matched_size
        @tree.children << Element.new(:math, @src.matched[1..-2], nil, category: :span, location: start_line_number)
    end

    define_parser(:inline_math, INLINE_MATH_START, '\$')
end
```
([GitHub Gist](https://gist.github.com/sp301415/db9f7bee03bda483aa2ca7c0ca92fbfa))

If you are using jekyll, you can put this ruby file under `_plugins` directory and set `input: GFMMKatex` under kramdown settings in `_config.yml`.

This overrides the default inline math parser, and parsers everything between two dollar signs. I tried getting `$$ ... $$` to work, and it kinda did if I just manually shoved math block element to `@tree.children`, but it rendered under `<p>` tag. (It looks just normal though.) Seems like there is no way to render it alongside `@tree`. 

Since it is a rather simple code (and just a ripoff of kramdown's own inline math parser tbf) it would be easy to modify to include other delimiters, etc etc. Hope this helps someone :)

**Comment by stsievert** [comment]
Thanks @sp301415! That's exactly what I'm looking for. I've tested it out, and your `GFMKatex` class works without modification and renders math as expected (without javascript!). Details on my Jekyll setup:

Relevant part of `_config.yml` (which is oriented for debugging):

``` yml
kramdown:
  input: GFMKatex  # GFMKatex class from sp301415 is in _plugins/server-side-math.rb (name not important, no modifications made)
  math_engine: sskatex
  math_engine_opts:
    katex_js: "js/katex/katex.min.js"
    debug: true
    verbose: true
    js_run: Node
    katex_opts:
      throwOnError: false
      macros: {}  # lots of macros here
```

My Gemfile:

``` 
source 'https://rubygems.org'

gem 'jekyll', '4.2.1'
gem 'kramdown-parser-gfm'
gem 'kramdown-math-sskatex'
gem 'kramdown', '~> 2.0'
gem 'execjs'
```

**Comment by leeway00** [comment]
@sp301415  @stsievert 
Thank you'all for the nice solution. 
I'm actually having an error with your code as below. I'm new to html and ruby so I'm not sure what I'm missing now. What should I check up to solve this error? I also put ```'kramdown-parser-gfm'``` in my Gemfile and GFMKatex in ```_plugins```


```
jekyll 3.9.0 | Error:  kramdown has no parser to handle the specified input format: GFMKatex
/Users/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/kramdown-2.3.1/lib/kramdown/document.rb:104:in `initialize': kramdown has no parser to handle the specified input format: GFMKatex (Kramdown::Error)
```

**Comment by sp301415** [comment]
@leeway00 I just checked in Jekyll 3.9.2, and I get the error too, albeit a different one (which seems to be irrelevant with my fix 🤷‍♂️): 
```
jekyll 3.9.0 | Error:  undefined method `valid_encoding?' for nil:NilClass

          if doc.content.valid_encoding?
                        ^^^^^^^^^^^^^^^^
```
I think it depends on Jekyll version; try upgrading it to at least 4.0.0. You must use GitHub Actions or other services like Netlify to deploy though, since GitHub Pages only supports Jekyll 3.9 and below out of the box.

**Comment by stsievert** [comment]
@sp301415 @leeway00 the first immediate thing I see is that you're both using Jekyll 3.9.0, not Jekyll 4.2.1. Does using Jekyll 4.2.1 make a difference? I think kramdown had some pretty significant changes since 2.0.

**Comment by sp301415** [comment]
@stsievert Yeah, seems like that fix only applies to jekyll >= 4.0.0.
